### ISSUE TITLE
Axis mismatch for output tensors in GPT module for TransFuser model

### ISSUE THREAD

**Issue by BXYMartin** [autonomousvision/transfuser#140]
Hi, thank you for the amazing work! When I am looking through the code for the GPT module in the TransFuser model, I found this part quite confusing to me. 

https://github.com/autonomousvision/transfuser/blob/a7d4db684c160095dec03851aff5ce92e36b2387/team_code_transfuser/transfuser.py#L361

So in my opinion, the tensor `x` here should be of size (batch_size, num_of_patches, num_of_channels). `x[:, :self.seq_len*self.img_vert_anchors*self.img_horz_anchors, :]` takes the corresponding image features from the tensor and then do the reshape with (batch_size, -1, image_height, image_width) where the `-1` is the feature channels. However when doing the reshape with `view`, the order of the axis should match or otherwise it is misaligned. The `.contiguous()` here removed the constraints on tensor shapes and made this part run without throwing an error. 

I have created a simple test to verify what I stated earlier:

```python
# Create a tensor of shape (batch_size, num_of_channel, image_height, image_width)
blob = torch.rand((5, 4, 3, 2))
# Now the tensor is of shape (batch_size, num_of_patches, num_of_channels)
patch = blob.view(5, 4, -1).permute(0, 2, 1)
# Check if they are identical
(patch.contiguous().view(5, 4, 3, 2) == blob).all()
# Or do a subtraction to check the values
patch.contiguous().view(5, 4, 3, 2) - blob
# Without .contiguous(), it will throw an exception
blob.view(5, 4, -1).permute(0, 2, 1).view(5, 4, 3, 2)
```

So in my opinion, I think `.permute()` is needed before reshaping the tensor. 

Thank you!

**Comment by Kait0** [comment]
That is a bug that unintentionally mixes up the channel and pixel dimensions.
Since the network can in principle learn to deal with it, and I'm not sure if it causes any problems, we left it in.
Changing it would break the pre-trained models.

As you mentioned, the problem is that at the point where the view gets applied, the pixel dimensions are still in front of the channel dimension due to the transformation earlier:
`image_tensor = image_tensor.view(bz, self.seq_len, -1, img_h, img_w).permute(0,1,3,4,2).contiguous().view(bz, -1, self.n_embd)`

I think the correct implementation should be:

```
image_tensor_out = x[:, :self.seq_len * self.config.img_vert_anchors * self.config.img_horz_anchors, :].view(
        bz * self.seq_len, img_h, img_w, -1).permute(0, 3, 1, 2).contiguous()
lidar_tensor_out = x[:, self.seq_len * self.config.img_vert_anchors * self.config.img_horz_anchors:, :].view(
          bz * self.seq_len, lidar_h, lidar_w, -1).permute(0, 3, 1, 2).contiguous()
```

**Comment by BXYMartin** [comment]
Yes, that explained my doubt, thank you so much!
